man/man1/snapcopy.1
.TH SNAPCOPY 1NEMO "12 February 2022"

.SH "NAME"
snapcopy - copy particles of snapshot subject to conditions

.SH "SYNOPSIS"
\fBsnapcopy in=\fIin_file\fP \fBout=\fIout_file\fP [parameter=value] .\|.\|.

.SH "DESCRIPTION"
\fIsnapcopy\fP copies an N-body system from \fBin\fP to \fBout\fP,
subjecting the individual particles to certain conditions.
.PP
It should be noted that there are a variety of other programs
which can perform copy functions under better circumstances.
\fIsnapsort\fP will rearrange the order of the particles in the
file, after which a number of them can be trimmed of
using either \fIsnapmask\fP or \fIsnapsample\fP. The program
\fIsnapplotedit\fP is an interactive editor, which allows 
the user to edit snapshots using polygonic areas in 2D projections
of 6D-phase space using the snapplot-interface. It requires a
special version of \fIyapp(3NEMO)\fP, which may not be implemented
on all systems.

.SH "PARAMETERS"
.so man1/parameters
.TP 24
\fBin=\fIin_file\fP
input file, in \fIsnapshot\fP(5NEMO) format.  [Default: none.]
.TP
\fBout=\fIout_file\fP
output file,  also \fIsnapshot\fP format.  [Default: none.]
.TP
\fBselect=\fIselect-string\fP
expression which evaluates to an integer, only particles with
non-zero expression will be copied to the output file. The ranking
of the particles is copied to the \fBKey\fP array in the snapshot
if not all particles were selected.
Selecting one body is cumbersome, e.g.
\fBselect=i==9\fP to select the 10th body, so a \fBibody=9\fP
a shortcut is provided for these case (see below).

The default is all particles will be copied.
[default: \fB1\fP].
.TP
\fBtimes=\fItime-ranges\fP
Only snapshots with time values within \fItimes-ranges\fP, which is
of the form \fBts[:te],...\fP, will be copied. [default: \fBall\fP].
.TP
\fBprecision=\fBdouble|single\fP
Precision of variables to be stored. The default for snapshots is
\fBdouble\fP (double precision), but many snapshot programs now
recognize data in single precision format. [default: \fBdouble\fP].
\fI *** This option is not implemented yet ***\fP
.TP
\fBkeep=\fIitems-list\fP
List of items to keep in the snapshot. The list may contain
any of \fBtime,mass,phase,phi,acc,aux\fP and \fBkey\fP.
.TP
\fBi=\fP
One body to select. If -1 is given, the \fBselect=\fP takes
preference. If a nunber 0 or larger is selected, this body is
selected. Only one body can be selected.   [-1]

.SH "SEE ALSO"
snapsort(1NEMO), snapmask(1NEMO), snapplotedit(5NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +3.0i
~/src/pjt/nbody   	snapcopy.1
.fi

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
12-apr-87	V1.0: document created          	PJT
26-sep-89	V1.1: debugged and exported to NEMO	PJT
12-feb-22	V1.4: implemented ibody=	PJT
.fi




src/nbody/trans/snapcopy.c
/*
 * SNAPCOPY.C: copy an N-body snapshot.
 *	
 *	11-Apr-89	V1.0	created 		PJT
 *	26-oct-90          a    ??
 *	11-jan-97	   b    report # copied, NEMOV2 style
 *	25-mar-97	   c    minor fix SINGLEPREC
 *       1-apr-01          d    compiler warning
 *      13-feb-04       V1.1f    silenced more compiler warnings (shetty bug?)
 *      15-nov-06        1.2    set time to 0 if it was absent     PJT/AP
 *    27-dec-2019        1.3    special case body= selection
 *
 *	BUG: should optionally copy other sets within the snapshot
 *	     set, e.g. diagnostics and story
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
				/* new filestruct */
#include <snapshot/snapshot.h>	
#include <snapshot/body.h>
#include <snapshot/get_snap.c>
#include <snapshot/put_snap.c>

#include <bodytransc.h>

string defv[] = {
    "in=???\n           Input snapshot",
    "out=???\n          Output snapshot",
    "select=1\n         Selection expression",
    "times=all\n        Times to select",
    "precision=double\n Precision of results to store (double/single) [unused]",
    "keep=all\n         Items to copy in snapshot",
    "i=-1\n             Select one body to select (overrides select=)",
    "VERSION=2.0\n      12-feb-2022 PJT",
    NULL,
};

string usage="copy an N-body snapshot";

string cvsid="$Id$";


#define TIMEFUZZ	0.0001	/* tolerance in time comparisons */

void nemo_main(void)
{
    stream instr, outstr;
    real   tsnap;
    string times, precision, keep;
    Body   *btab = NULL, *bpi, *bpo;
    int    i, nbody, nout, nreject, bitsi, bitso, vis, visnow, vismax;
    int    isnap = 0;
    bool   Qall;
    int    ibody = getiparam("i");
    iproc_body sfunc;

    if (ibody >=0) warning("ibody not implemented yet");

    times = getparam("times");
    sfunc = btitrans(getparam("select"));
    instr = stropen(getparam("in"), "r");
    outstr = stropen(getparam("out"), "w");
    precision = getparam("precision");	/* unused */
    if (!streq(precision,"double")) {
        warning("Precision \"%s\" not supported yet, use csf convert=d2f/f2d",
            precision);
    }
    keep = getparam("keep");
    bitso = 0;
    if (scanopt(keep,"all")) {
        Qall = TRUE;
    } else {
        Qall = FALSE;
        if (scanopt(keep,"mass"))
            bitso |= MassBit;
        if (scanopt(keep,"phase"))
            bitso |= PhaseSpaceBit;
        if (scanopt(keep,"phi"))
            bitso |= PotentialBit;
        if (scanopt(keep,"acc"))
            bitso |= AccelerationBit;
        if (scanopt(keep,"aux"))
            bitso |= AuxBit;
        if (scanopt(keep,"key"))
            bitso |= KeyBit;
        if (scanopt(keep,"time"))
            bitso |= TimeBit;
    }


    get_history(instr);
    put_history(outstr);		
    for (;;) {
    	get_history(instr);		/* skip over stuff we can forget */
        if (!get_tag_ok(instr, SnapShotTag))
		break;			/* done with work in loop */
        get_snap(instr, &btab, &nbody, &tsnap, &bitsi);
        if ((bitsi & MassBit) == 0 && (bitsi & PhaseSpaceBit) == 0) {
	    continue;       /* just skip */
        }
        				/* do some initial output */
        if ((bitsi & TimeBit) == 0) {	/* if not time flag */
        	tsnap = 0.0;		/* set time to zero */
		bitsi |= TimeBit;
		warning("time reset to 0");
        } else if (!streq(times,"all") && !within(tsnap, times, TIMEFUZZ))
        	continue;		/* now skip this snapshot */
        if (bitsi & KeyBit)
            dprintf(0,"Warning: Keyfield reinitialized\n");
        for (bpi = btab; bpi < btab+nbody; bpi++)
            Key(bpi) = 0;                    /* set to false */
        visnow = vismax = nreject = 0;
	if (ibody < 0) {   // use the select=
	  do {				/* loop through all particles */
	    visnow++;
            for (bpi = btab, i=0; i<nbody; bpi++,i++) {
	      vis = (*sfunc)(bpi, tsnap, i);
	      dprintf(2,"sfunc [%d] = %d\n",i,vis);
	      vismax = MAX(vismax,vis);
	      if (vis==visnow)
		Key(bpi) = 1;
            }
	  } while (visnow < vismax);          /* do all layers */
	} else
	  Key(btab+ibody) = 1;
        nreject = 0;
        for (bpi = btab, bpo = btab, i=0; i<nbody; bpi++,i++) {
            if (!Key(bpi)) { 
               nreject++;       /* count number of stars rejected */
               continue;
            }
            if (nreject)        /* only copy while out of sync */
                bcopy (bpi, bpo, sizeof(Body));
            Key(bpo) = i;       /* counting from zero */
            bpo++;
            
        }
        nout = nbody - nreject;
        if (nout) {
            if (Qall) {             /* if all old things selected */
                if (nreject)        /* and some stars were rejected */
                    bitsi |= KeyBit;    /* then explicitly add Key field */
            } else
                bitsi = bitso;      
            put_snap(outstr, &btab, &nout, &tsnap, &bitsi);
            dprintf (1,"Snapshot time=%f copied %d particles\n",
				tsnap,nout);
        } else
           dprintf(0,"No particles to copy at tsnap=%f\n",tsnap);
	isnap++;
	progress(1.0,"Processed snapshot %d", isnap);
    }
}
